core/services/mongodb-local/src/main/java/com/mongodb/stitch/core/services/mongodb/local/internal/EmbeddedMongoClientFactory.java

```diff
@@ -17,57 +17,78 @@
 package com.mongodb.stitch.core.services.mongodb.local.internal;
 
 import com.mongodb.client.MongoClient;
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
 import org.bson.codecs.configuration.CodecRegistry;
 
 public abstract class EmbeddedMongoClientFactory {
-
-  private static final Map<String, MongoClient> instances = new ConcurrentHashMap<>();
+  private static final MongoClientHolder clientHolder = new MongoClientHolder();
 
   protected abstract MongoClient createClient(
       final String dbPath,
       final CodecRegistry codecRegistry);
 
   protected EmbeddedMongoClientFactory() {
     Runtime.getRuntime().addShutdownHook(new Thread(() -> {
-      for (final MongoClient client : instances.values()) {
-        client.close();
+      if (clientHolder.client != null) {
+        clientHolder.client.close();
       }
     }));
   }
 
-  public synchronized MongoClient getClient(
+  public MongoClient getClient(
       final String key,
       final String dbPath,
       final CodecRegistry codecRegistry
   ) {
+    if (key == null || key.isEmpty()) {
+      throw new IllegalArgumentException("keymust be non-empty");
+    }
     if (dbPath == null || dbPath.isEmpty()) {
       throw new IllegalArgumentException("dbPath must be non-empty");
     }
-    if (instances.containsKey(key)) {
-      return instances.get(key);
-    }
 
-    for (final MongoClient value : instances.values()) {
-      value.close();
```

Comment:
Seems like strange repetition below with the same if statements. I feel like you should be able to take out the first if statement, but I am not as familiar with concurrent Java programming

Reply:
Example:
```1:   if (someCondition) {
2:        synchronized(someObject) {
3:            doSomething();
4:        }
5:    }

Suppose that two threads hit line 1 at the same time, they will both evaluate the same result for `someCondition`. Now supposing `someCondition` is true, they will both attempt to lock on `someObject`.  One will succeed and proceed to perform `doSomething()` while the other blocks and waits for the lock on `someObject` to become available. When the winning thread exits the `synchronized` block the losing thread will then enter it and also perform `doSomething()`

If `doSomething` is idempotent then it doesn't matter, but if `doSomething` has side effects then this is broken. The double-check pattern, though imperfect, prevents the losing thread from duplicating the inner operation.

Reply:
I understand why you need the inner condition, but I was more confused about why you would need the outer one. I guess in the case that someCondition is false you avoid entering the synchronized block at all?